dspace-api/src/main/java/org/dspace/curate/WorkflowCuratorServiceImpl.java

```diff
@@ -118,14 +124,23 @@ public boolean doCuration(Context c, BasicWorkflowItem wfi)
             Curator curator = new Curator();
             // are we going to perform, or just put on queue?
             if (step.queue != null) {
+                // The queue runner will call setReporter
                 for (Task task : step.tasks) {
                     curator.addTask(task.name);
                 }
                 curator.queue(c, String.valueOf(wfi.getID()), step.queue);
                 basicWorkflowItemService.update(c, wfi);
                 return false;
             } else {
-                return curate(curator, c, wfi);
+                Date now = GregorianCalendar.getInstance().getTime();
+                SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMddThhmmssSSS");
+                String filename = String.format("curation-%s.log", sdf.format(now));
+                Path logPath = Paths.get(
+                        configurationService.getProperty("dspace.dir"), "logs", filename);
+                try (PrintWriter reporter = new PrintWriter(logPath.toFile())) {
```

Comment:
Why are we "manually" managing this log file? Is there a way to use log4j for this?  It seems a bit odd to me to do this outside of our `log4j.properties` file, and hardcode it to a specific name/date format.  Ideally, I'd want to manage this alongside all other DSpace logs.

Reply:
This code yields a separate log file for each run; log4j would collect them all in daily aggregate files.  I'd be happy to discuss how we want to capture the output of curation runs that happen without an admin present.

I should bear in mind that WorkflowCuratorService would always be applied to single Items.  OTOH there is no limit to how much a task may report for a single object.

Reply:
As a sidenote... I *believe* some amount of curation task information already is "discovered" by log4j and ends up in the `dspace.log` file.  I know I've seen curation tasks trigger output in the `dspace.log` before.  So, if we are now thinking a separate curation.log file is useful (and in my opinion it seems useful), we may want to redirect that output as well.

Reply:
I do wonder whether we want to distinguish between logging stuff that the code mutters about its own operation, and tasks' reports on the objects on which they operate.  Something about the word "report" makes me think it should be separate.  Maybe I should have directed the files to [DSpace]/reports.  (Come to think of it, if this code continues to make its own files, the destination should be configurable.)

The main dspace.log is such a rag-bag of unconnected stuff that I often ponder how to break it apart into multiple streams with more focus.

Anyway it should not be difficult to whip up an Appendable implementation that just stuffs the characters into a log category.  Perhaps a specific, documented category that can be directed to a separate log appender.  "workflow.${collection UUID}.${taskID}" or some such.

I wonder if it is worth bothering to worry about what happens when two workflows are progressing in parallel -- what would that look like, coming out a single reporting channel?  Worst case would be two (or more!) GUI batch loads with tasks attached to Step 1, so we have multiple curations per second going off immediately on submission.  This is why the separate report filenames are dated down to the millisecond....

Reply:
Aside:  I've come to think that the CLI and the webapp.s should each have its own log configuration and log files.

Reply:
If you want reports to the millisecond, and want separate tasks in separate logs, I'd assume you can configure that in log4j.  In log4j you can configure the log file naming structure...you can also configure which individual Java packages/classes log to which log files.

I'm just worried here about generating unexpected logs outside of log4j. Plus, the lack of configurability here could become a pain point to some, based on how they manage their logs (as some folks do use log analyzers to "watch" for errors/warnings and unexpected logs using a hardcoded naming convention/path can mean information is overlooked or missed)

Reply:
@terrywbrady seems to have some specific ideas about how task reports would be used.  Maybe he can help us sort this out.

Reply:
I would say that errors and warnings should go to dspace.log, just as with the rest of the code.  I wouldn't use the report stream for that.  I'd use reporting to say "I successfully operated on object X and here's what I have to say as a result."

Reply:
We could punt:  define interface o.d.curation.ReportWriter, and 2-3 stock implementation plugins.  A LogReportWriter, a FileReportWriter, maybe a CommandReportWriter....

Reply:
@mwoodiupui : It's worth mentioning that (as of DSpace 6.x) the location of DSpace log files is actually also configurable within log4j (and *is not controlled by the dspace.cfg in any way*):
https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace/config/log4j.properties#L19
(This change came in as part of the new Configuration System, and was somewhat necessitated by it -- as log4j could not "sync up" with Apache Commons Configuration)

This `log.dir` in log4j is autopopulated during `ant` deployment, but could obviously be modified after deployment as needed.

So, if an institution changes their log4j configuration to store the DSpace logs under `/var/log/dspace/` (or wherever), this _should_ work fine....except that, hardcoded log locations obviously won't be aware of that, and will continue to log elsewhere.

This is why I'm pushing a bit hard on using log4j.  If we don't use log4j, then logs could end up in unexpected locations, since all logs are now configured via log4j.

Reply:
Ah, that's a plain bug.  I didn't remember `log.dir` or I should have used it.  But I think it was a mistake to use the 'log' directory at all for files I don't think of as "logs".

Anyway I think we need a few more discussants to work out how reports are going to be used and how they ought to be organized, and then see how to put them in the right place by the right path.

Reply:
@mwoodiupui , here are my thoughts about task reports.

I would like to create Curation Tasks that enforce expectations for the contents of a collection such as the relationships between bitstreams and metadata.

For instance, I would like to scan a community and verify that every item has a type and that type matches a value from a list of standard types.  Based on the type of an item, I would then like to ensure that the bitstreams conform to my expectations.  For instance, if I have a type of "Thesis", I would expect to see a PDF, a thumbnail, and some extracted text.  

For items with a type of "Audio" or "Video", I would have a different set of standards to apply.

For any items that do not conform to these rules, I would like to report (1)a URL to the item in question and (2)a description of the non-conforming attributes of the item.

Because a curation task can be run in the background, it would be possible to set up a robust set of checks that would be run against each collection a few times per year.

Reply:
Over the weekend, I sketched up untested LogReporter and FileReporter plugins together with the plugin plumbing.  They aren't hard to write, and now there is a Reporter interface (needed by the PluginService) for a guide.  I haven't pushed the changes up yet, awaiting discussion to see what is really needed.

The FileReporter writes per-run files to [DSpace]/reports (configurable).  The LogReporter writes whole lines (strings ending in newline) to log records through the category "curation.$taskID" so they can be picked off to a separate curation log or even separate logs per taskID.